### PR TITLE
Plugin reload <plugin_name> feature

### PR DESCRIPTION
**Pull request by pwnorbitals**
First PR here, please give feedback
Ref to #365

### CODE REVIEW

**Comment by 12xx12** [comment]
For the checkstyle the lua plugin is included in src/CheckBasicStyle.lua

**Comment by pwnorbitals** [comment]
Wow, those requirements are stringent - I’ll see later on if I can make autoformatting rules, at least for the whitespaces.

**Comment by 12xx12** [comment]
Lua API documentation goes into Server/Plugins/APIDump/

Default goes to APIDesc.lua

Some stuff goes into the subfolders.

If you want to check run

```
load APIDump
api
```

In the console of cuberite

**Comment by pwnorbitals** [comment]
I added to Classes/Plugins.lua, not sure if I should edit APIDesc.lua, it’s a big file it’s difficult to know what to edit

**Comment by 12xx12** [comment]
The output of ```api``` is a bit broken - yeah this error is there since icontribute here :D

The two missing symbols also go into Classes/Plugins.lua

I use the search feature of my IDE
The main principle is that you should keep everything alpha-sorted. From these .lua files the plugin generates the HTML files you can see on https://api.cuberite.org/

**Comment by 12xx12** [comment]
oh - and add yourself to the CONTRIBUTING file

this is for legal reason. if you have questions @bearbin may help you with that

File: CONTRIBUTORS [hunk 1]
```diff
@@ -76,3 +76,4 @@ worktycho
 Xenoxis
 xoft (Mattes Dolak/madmaxoft on GH)
 Yeeeeezus (Donated AlchemistVillage prefabs)
+pwnOrbitals
```
Comment (12xx12) [inline CONTRIBUTORS:79]: Newline at EOF

File: src/Bindings/PluginManager.cpp [hunk 2]
```diff
@@ -166,34 +166,51 @@ void cPluginManager::InsertDefaultPlugins(cSettingsRepositoryInterface & a_Setti
...
 	}
-	for (auto & folder: PluginsToUnload)
+
+
+	for (auto & currentPlugin: PluginsNeedAction)
 	{
-		bool HasUnloaded = false;
-		bool HasFound = false;
+		auto & action = currentPlugin.first;
+		auto & folder = currentPlugin.second;
```
Comment (12xx12) [inline src/Bindings/PluginManager.cpp:179]:
```suggestion
		auto & Action = CurrentPlugin.first;
		auto & Folder = CurrentPlugin.second;
```

File: src/Bindings/PluginManager.cpp [hunk 3]
```diff
@@ -166,34 +166,51 @@ void cPluginManager::InsertDefaultPlugins(cSettingsRepositoryInterface & a_Setti
...
 	{
-		cCSLock Lock(m_CSPluginsToUnload);
-		std::swap(m_PluginsToUnload, PluginsToUnload);
+		cCSLock Lock(m_CSPluginsNeedAction);
+		std::swap(m_PluginsNeedAction, PluginsNeedAction);
 	}
-	for (auto & folder: PluginsToUnload)
+
+
+	for (auto & currentPlugin: PluginsNeedAction)
```
Comment (12xx12) [inline src/Bindings/PluginManager.cpp:176]:
```suggestion
	for (auto & CurrentPlugin: PluginsNeedAction)
```
We use CamelCase for local variable names.

File: src/Bindings/PluginManager.cpp [hunk 4]
```diff
@@ -166,34 +166,51 @@ void cPluginManager::InsertDefaultPlugins(cSettingsRepositoryInterface & a_Setti
...
 	{
-		bool HasUnloaded = false;
-		bool HasFound = false;
+		auto & action = currentPlugin.first;
+		auto & folder = currentPlugin.second;
+
+		bool WasLoaded = false;
+		bool WasFound = false;
 		for (auto & plugin: m_Plugins)
```
Comment (12xx12) [inline src/Bindings/PluginManager.cpp:183]:
```suggestion
		for (auto & Plugin: m_Plugins)
```

File: src/Bindings/Plugin.cpp [hunk 5]
```diff
@@ -43,6 +43,16 @@ void cPlugin::Unload(void)
...
+	Load();
+}
+
+
+
+
+
 AString cPlugin::GetLocalFolder(void) const
 {
 	return std::string("Plugins/") + m_FolderName;
```
Comment (12xx12) [inline src/Bindings/Plugin.cpp:58]: i know this is not your business but can you do
```suggestion
	return AString("Plugins") + cFile::GetPathSeparator() + m_FolderName;
```

Comment (12xx12) [reply src/Bindings/Plugin.cpp:58]: I'm just wondering why this didn't break in windows OS yet

**Review by 12xx12 — CHANGES_REQUESTED** [review]

File: CONTRIBUTORS (hunk 1 above)
Comment (bearbin) [inline CONTRIBUTORS:79]: This should be alpha-sorted.

File: src/Server.cpp [hunk 6]
```diff
@@ -463,7 +463,15 @@ void cServer::ExecuteConsoleCommand(const AString & a_Cmd, cCommandOutputCallbac
 	}
 	else if (split[0] == "reload")
 	{
-		cPluginManager::Get()->ReloadPlugins();
+		if (split.size() > 1)
+		{
+			cPluginManager::Get()->ReloadPlugin(split[1]);
+			a_Output.Out("Plugin unloaded scheduled");
```
Comment (bearbin) [inline src/Server.cpp:469]: This looks like a typo - it should be "Plugin unload scheduled"

**Review by bearbin — CHANGES_REQUESTED** [review]
Just two tiny changes to make, this looks like a really great first PR.

**Comment by pwnorbitals** [comment]
> Just two tiny changes to make, this looks like a really great first PR.

Thanks, it means a lot to me. Looking forward to contribute some more  :)

**Review by bearbin — APPROVED** [review]

**Pull request merged**